mne/io/base.py

```diff
@@ -836,6 +836,72 @@ def __setitem__(self, item, value):
         # set the data
         self._data[sel, start:stop] = value
 
+    def _get_data(self, sel, start, stop):
+        """Get data omitting bad segments."""
+        if self.annotations is None:
+            return self[sel, start:stop]
+        sfreq = self.info['sfreq']
+        bads = [idx for idx, desc in enumerate(self.annotations.description)
+                if desc.upper().startswith('BAD')]
+        onsets = self.annotations.onset[bads]
+        onsets = _onset_to_seconds(self, onsets)
+        ends = onsets + self.annotations.duration[bads]
+        omit = np.concatenate([np.where(onsets > stop / sfreq)[0],
+                               np.where(ends < start / sfreq)[0]])
+        onsets, ends = np.delete(onsets, omit), np.delete(ends, omit)
+        if len(onsets) == 0:
+            return self[sel, start:stop]
+        stop = min(stop, self.n_times)
+        order = np.argsort(onsets)
+        onsets = self.time_as_index(onsets[order])
+        ends = self.time_as_index(ends[order])
+
+        # Merge overlaps.
```

Comment:
Would the next 30 lines be equivalently simpler as the following 10:
```
np.clip(onsets, start, stop, onsets)
np.clip(ends, start, stop, ends)
used = np.ones(stop - start, bool)
for onset, end in zip(onsets, ends):
    if onset >= end:
        continue
    used[onset - start: end - start] = True
used = np.concatenate([[False], used, [False]])
starts = np.where(~used[:-1] & used[1:])[0] + start
stops = np.where(used[:-1] & ~used[1:])[0] + start
```
? Maybe I'm missing something, though.